beacon-chain/core/state/transition.go

```diff
@@ -100,7 +106,211 @@ func ProcessBlock(state *pb.BeaconState, block *pb.BeaconBlock) (*pb.BeaconState
+	if err != nil {
+		return nil, fmt.Errorf("could not get current boundary attestations: %v", err)
+	}
+	currentBoundaryAttesterIndices, err := v.ValidatorIndices(state, currentBoundaryAttestations)
+	if err != nil {
+		return nil, fmt.Errorf("could not get current boundary attester indices: %v", err)
+	}
+	currentBoundaryAttestingBalances := e.TotalBalance(state, currentBoundaryAttesterIndices)
+
+	// Calculate the attesting balances of validators that made an attestation
+	// during previous epoch.
+	prevAttestations := e.PrevAttestations(state)
+	prevAttesterIndices, err := v.ValidatorIndices(state, prevAttestations)
+	if err != nil {
+		return nil, fmt.Errorf("could not get prev epoch attester indices: %v", err)
+	}
+
+	// Calculate the attesting balances of validators that targeted
+	// previous justified hash.
+	prevJustifiedAttestations := e.PrevJustifiedAttestations(state,
+		currentAttestations, prevAttestations)
+	prevJustifiedAttesterIndices, err := v.ValidatorIndices(state, prevJustifiedAttestations)
+	if err != nil {
+		return nil, fmt.Errorf("could not get prev epoch justified attester indices: %v", err)
+	}
+	prevJustifiedAttestingBalance := e.TotalBalance(state, prevJustifiedAttesterIndices)
+
+	// Calculate the attesting balances of validator justifying epoch boundary block
+	// at the start of previous epoch.
+	prevBoundaryAttestations, err := e.PrevBoundaryAttestations(state, prevJustifiedAttestations)
+	if err != nil {
+		return nil, fmt.Errorf("could not get prev boundary attestations: %v", err)
+	}
+	prevBoundaryAttesterIndices, err := v.ValidatorIndices(state, prevBoundaryAttestations)
+	if err != nil {
+		return nil, fmt.Errorf("could not get prev boundary attester indices: %v", err)
+	}
+	prevBoundaryAttestingBalances := e.TotalBalance(state, prevBoundaryAttesterIndices)
+
+	// Calculate attesting balances of validator attesting to expected beacon chain head
+	// during previous epoch.
+	prevHeadAttestations, err := e.PrevHeadAttestations(state, prevAttestations)
+	if err != nil {
+		return nil, fmt.Errorf("could not get prev head attestations: %v", err)
+	}
+	prevHeadAttesterIndices, err := v.ValidatorIndices(state, prevHeadAttestations)
+	if err != nil {
+		return nil, fmt.Errorf("could not get prev head attester indices: %v", err)
+	}
+	prevHeadAttestingBalances := e.TotalBalance(state, prevHeadAttesterIndices)
+
+	// Process receipt roots.
+	if e.CanProcessReceiptRoots(state) {
+		e.ProcessReceipt(state)
+	}
+
+	// Update justification.
+	state = e.ProcessJustification(
+		state,
+		currentBoundaryAttestingBalances,
+		prevBoundaryAttestingBalances,
+		totalBalance)
+
+	// Update Finalization.
+	state = e.ProcessFinalization(state)
+
+	// Process crosslinks records.
+	state, err = e.ProcessCrosslinks(
+		state,
+		currentAttestations,
+		prevAttestations)
+	if err != nil {
+		return nil, fmt.Errorf("could not process crosslink records: %v", err)
+	}
+
+	// Process casper rewards and penalties.
```

Comment:
https://github.com/ethereum/eth2.0-specs/blob/master/specs/core/0_beacon-chain.md#rewards-and-penalties